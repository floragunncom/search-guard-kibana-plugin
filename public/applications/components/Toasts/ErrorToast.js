/*
 *    Copyright 2020 floragunn GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { get } from 'lodash';
import { EuiText, EuiButton, EuiFlexGroup, EuiFlexItem, EuiSpacer } from '@elastic/eui';

export function ErrorToast({ error, errorMessage, errorDetails, onDetailsClick }) {
  errorMessage = errorMessage || error.message;
  errorDetails = errorDetails || get(error, 'body', undefined);

  try {
    errorDetails = JSON.stringify(errorDetails, null, 2);
  } catch (e) {
    console.error('addErrorToast', 'Failed to stringify the error details.', e);
    errorDetails = null;
  }

  if (errorDetails) {
    return (
      <div id="sgErrorToast">
        <EuiText size="s">
          <p>{errorMessage}</p>
        </EuiText>
        <EuiSpacer />

        <EuiFlexGroup justifyContent="flexEnd">
          <EuiFlexItem grow={false}>
            <EuiButton
              color="danger"
              size="s"
              onClick={() => onDetailsClick({ errorMessage, errorDetails })}
            >
              Details
            </EuiButton>
          </EuiFlexItem>
        </EuiFlexGroup>
      </div>
    );
  }

  return <div id="sgErrorToast">{errorMessage}</div>;
}

ErrorToast.propTypes = {
  error: PropTypes.object.isRequired,
  onDetailsClick: PropTypes.func,
  errorMessage: PropTypes.string,
  errorDetails: PropTypes.object,
};
