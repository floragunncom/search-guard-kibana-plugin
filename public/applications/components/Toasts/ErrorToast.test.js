/*
 *    Copyright 2020 floragunn GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { ErrorToast } from './ErrorToast';

describe(ErrorToast.name, () => {
  it('renders correctly', () => {
    const error = new Error('nasty!');

    const tree = renderer.create(<ErrorToast error={error} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders details correctly', () => {
    const error = new Error('nasty!');
    error.body = {
      attributes: {
        body: {
          a: [0, 1],
          b: { c: 2 },
        },
      },
    };

    const tree = renderer.create(<ErrorToast error={error} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly if details cannot be parsed', () => {
    const error = new Error('nasty!');
    const a = { next: null };
    const b = { next: null };
    a.next = b;
    b.next = a;
    error.body = { attributes: { body: { a } } };

    const tree = renderer.create(<ErrorToast error={error} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly if custom message and details', () => {
    const error = new Error('nasty!');
    const errorMessage = 'Super nasty!!!';
    const errorDetails = {
      attributes: {
        body: {
          a: [0, 1],
          b: { c: 2 },
        },
      },
    };

    const tree = renderer
      .create(<ErrorToast error={error} errorMessage={errorMessage} errorDetails={errorDetails} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
