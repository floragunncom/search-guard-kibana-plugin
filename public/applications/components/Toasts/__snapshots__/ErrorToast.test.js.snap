// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorToast renders correctly 1`] = `
<div
  id="sgErrorToast"
>
  nasty!
</div>
`;

exports[`ErrorToast renders correctly if custom message and details 1`] = `
<div
  id="sgErrorToast"
>
  <div
    className="euiText euiText--small"
  >
    <p>
      Super nasty!!!
    </p>
  </div>
  <div
    className="euiSpacer euiSpacer--l"
  />
  <div
    className="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--justifyContentFlexEnd euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      className="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <button
        className="euiButton euiButton--danger euiButton--small"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <span
          className="euiButtonContent euiButton__content"
        >
          <span
            className="euiButton__text"
          >
            Details
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`ErrorToast renders correctly if details cannot be parsed 1`] = `
<div
  id="sgErrorToast"
>
  nasty!
</div>
`;

exports[`ErrorToast renders details correctly 1`] = `
<div
  id="sgErrorToast"
>
  <div
    className="euiText euiText--small"
  >
    <p>
      nasty!
    </p>
  </div>
  <div
    className="euiSpacer euiSpacer--l"
  />
  <div
    className="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--justifyContentFlexEnd euiFlexGroup--directionRow euiFlexGroup--responsive"
  >
    <div
      className="euiFlexItem euiFlexItem--flexGrowZero"
    >
      <button
        className="euiButton euiButton--danger euiButton--small"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <span
          className="euiButtonContent euiButton__content"
        >
          <span
            className="euiButton__text"
          >
            Details
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;
